src/goto-harness/goto_harness_parse_options.cpp

```diff
@@ -7,46 +7,168 @@ Author: Diffblue Ltd.
 \******************************************************************/
 
 #include <cstddef>
-#include <iostream>
+#include <set>
 #include <string>
+#include <utility>
 
+#include <goto-programs/goto_convert.h>
+#include <goto-programs/goto_model.h>
+#include <goto-programs/read_goto_binary.h>
+#include <goto-programs/write_goto_binary.h>
+#include <util/exception_utils.h>
 #include <util/exit_codes.h>
+#include <util/invariant.h>
 #include <util/version.h>
 
+#include "function_harness_generator.h"
+#include "goto_harness_generator_factory.h"
 #include "goto_harness_parse_options.h"
 
+// The basic idea is that this module is handling the following
+// sequence of events:
+// 1. Initialise a goto-model by parsing an input (goto) binary
+// 2. Initialise the harness generator (with some config) that will handle
+//    the mutation of the goto-model.
+// 3. Write the end result of that process to the output binary
+
 int goto_harness_parse_optionst::doit()
 {
   if(cmdline.isset("version"))
   {
-    std::cout << CBMC_VERSION << '\n';
+    status() << CBMC_VERSION << '\n';
     return CPROVER_EXIT_SUCCESS;
   }
 
-  help();
-  return CPROVER_EXIT_USAGE_ERROR;
+  auto config = handle_common_options();
+  auto factory = make_factory();
+
+  auto factory_options = collect_generate_factory_options();
+
+  // Initialise harness generator
+  auto harness_generator =
+    factory.factory(config.harness_type, factory_options);
+  CHECK_RETURN(harness_generator != nullptr);
+
+  // Read goto binary into goto-model
+  auto read_goto_binary_result =
+    read_goto_binary(config.in_file, get_message_handler());
+  if(!read_goto_binary_result.has_value())
+  {
+    throw deserialization_exceptiont{"failed to read goto program from file `" +
+                                     config.in_file + "'"};
+  }
+  auto goto_model = std::move(read_goto_binary_result.value());
```

Comment:
You will likely also need to do `config.set_from_symbol_table(goto_model.symbol_table);` at which point you will want to rename the local `config` variable.